based on https://github.com/ungeskriptet/linux/blob/sc7180-ufs/arch/arm64/boot/dts/qcom/sc7180.dtsi#L1464-L1530
and https://github.com/ungeskriptet/linux/commit/4e9939bf0d53fbe9820053ba0a1f8b8f2cdcec77

credits go to ungeskriptet - see also: https://oftc.irclog.whitequark.org/aarch64-laptops/2023-06-30#32270163
and from there on the following days

diff --git a/arch/arm64/boot/dts/qcom/sc7180.dtsi b/arch/arm64/boot/dts/qcom/sc7180.dtsi
index a65be760d1a7..31a6a0b1c8cd 100644
--- a/arch/arm64/boot/dts/qcom/sc7180.dtsi
+++ b/arch/arm64/boot/dts/qcom/sc7180.dtsi
@@ -1461,6 +1524,74 @@ mmss_noc: interconnect@1740000 {
 			qcom,bcm-voters = <&apps_bcm_voter>;
 		};
 
+                ufs_mem_hc: ufshc@1d84000 {
+                        compatible = "qcom,sc7180-ufshc", "qcom,ufshc",
+                                     "jedec,ufs-2.0";
+                        reg = <0 0x01d84000 0 0x3000>,
+                              <0 0x01d90000 0 0x8000>;
+                        reg-names = "std", "ice";
+                        interrupts = <GIC_SPI 265 IRQ_TYPE_LEVEL_HIGH>;
+                        phys = <&ufs_mem_phy>;
+                        phy-names = "ufsphy";
+                        lanes-per-direction = <1>;
+                        #reset-cells = <1>;
+                        resets = <&gcc GCC_UFS_PHY_BCR>;
+                        reset-names = "rst";
+
+                        iommus = <&apps_smmu 0xa0 0>;
+
+                        clock-names =
+                                "core_clk",
+                                "bus_aggr_clk",
+                                "iface_clk",
+                                "core_clk_unipro",
+                                "ref_clk",
+                                "tx_lane0_sync_clk",
+                                "rx_lane0_sync_clk",
+                                "ice_core_clk";
+                        clocks =
+                                <&gcc GCC_UFS_PHY_AXI_CLK>,
+                                <&gcc GCC_AGGRE_UFS_PHY_AXI_CLK>,
+                                <&gcc GCC_UFS_PHY_AHB_CLK>,
+                                <&gcc GCC_UFS_PHY_UNIPRO_CORE_CLK>,
+                                <&rpmhcc RPMH_CXO_CLK>,
+                                <&gcc GCC_UFS_PHY_TX_SYMBOL_0_CLK>,
+                                <&gcc GCC_UFS_PHY_RX_SYMBOL_0_CLK>,
+                                <&gcc GCC_UFS_PHY_ICE_CORE_CLK>;
+                        freq-table-hz =
+                                <50000000 200000000>,
+                                <0 0>,
+                                <0 0>,
+                                <37500000 150000000>,
+                                <0 0>,
+                                <0 0>,
+                                <0 0>,
+                                <0 300000000>;
+
+                        status = "disabled";
+                };
+
+                ufs_mem_phy: phy@1d87000 {
+                        compatible = "qcom,sc7180-qmp-ufs-phy";
+                        reg = <0 0x01d87000 0 0x1000>;
+                        #address-cells = <2>;
+                        #size-cells = <2>;
+                        ranges;
+                        clock-names = "ref",
+                                      "ref_aux";
+                        clocks = <&gcc GCC_UFS_MEM_CLKREF_CLK>,
+                                 <&gcc GCC_UFS_PHY_PHY_AUX_CLK>;
+
+                        power-domains = <&gcc UFS_PHY_GDSC>;
+
+                        resets = <&ufs_mem_hc 0>;
+                        reset-names = "ufsphy";
+
+                        #phy-cells = <0>;
+
+                        status = "disabled";
+                };
+
 		ipa: ipa@1e40000 {
 			compatible = "qcom,sc7180-ipa";
 
diff --git a/drivers/phy/qualcomm/phy-qcom-qmp-ufs.c b/drivers/phy/qualcomm/phy-qcom-qmp-ufs.c
index 8c877b668bb9..c686c845dd40 100644
--- a/drivers/phy/qualcomm/phy-qcom-qmp-ufs.c
+++ b/drivers/phy/qualcomm/phy-qcom-qmp-ufs.c
@@ -1637,6 +1637,9 @@ static const struct of_device_id qmp_ufs_of_match_table[] = {
 	}, {
 		.compatible = "qcom,sa8775p-qmp-ufs-phy",
 		.data = &sa8775p_ufsphy_cfg,
+	}, {
+		.compatible = "qcom,sc7180-qmp-ufs-phy",
+		.data = &sm7150_ufsphy_cfg,
 	}, {
 		.compatible = "qcom,sc8180x-qmp-ufs-phy",
 		.data = &sm8150_ufsphy_cfg,
